Bump Go version and update dependencies

valaparthvi opened the pull request:
### What does this PR do?
1. Update to Go 1.20
2. Update `onsi/ginkgo`, `onsi/gomega`, and other non-k8s pkgs to `latest`.
3. Update `k8s.io/api`, `k8s.io/apimachinery`, `rancher/client-go` to support k8s 1.28.
4. ~Update rancher/rancher to the HEAD commit of branch [release/v2.8.1](https://github.com/rancher/rancher/tree/release/v2.8.1).~
5. Move from rancher/rancher to rancher/shepherd.

This dependency update is targeted towards moving to Go 1.20 and migrating to rancher/shepherd from rancher/rancher.
 
### Which issue(s) this PR fixes (optional, in fixes #<issue number>(, fixes #<issue_number>, ...) format, will close the issue(s) when PR gets merged):
Fixes #44 

### Checklist:
- [x] Squashed commits into logical changes
- [ ] Documentation
- [x] GitHub Actions (if applicable)
    1. [~GKE-E2E~](https://github.com/rancher/hosted-providers-e2e/actions/runs/7692136164)
    2. [~EKS-E2E~](https://github.com/rancher/hosted-providers-e2e/actions/runs/7692143244)
    6. [~AKS-E2E~](https://github.com/rancher/hosted-providers-e2e/actions/runs/7692134023)
    7. [~AKS-E2E-2~](https://github.com/rancher/hosted-providers-e2e/actions/runs/7724184585)
    8. [~EKS-E2E-2~](https://github.com/rancher/hosted-providers-e2e/actions/runs/7724240711)
    9. [~GKE-E2E-2~](https://github.com/rancher/hosted-providers-e2e/actions/runs/7728264229)
    10. ~[GKE-E2E-3](https://github.com/rancher/hosted-providers-e2e/actions/runs/7754756725)~
    11. ~[EKS-E2E-3](https://github.com/rancher/hosted-providers-e2e/actions/runs/7754758887)~
    12. ~[AKS-E2E-3](https://github.com/rancher/hosted-providers-e2e/actions/runs/7754759939)~
    13. [AKS-E2E-4](https://github.com/rancher/hosted-providers-e2e/actions/runs/7780938596)
    14. [EKS-E2E-4](https://github.com/rancher/hosted-providers-e2e/actions/runs/7780939567)
    15. [GKE-E2E-4](https://github.com/rancher/hosted-providers-e2e/actions/runs/7780940722)

### Special notes for your reviewer:
<details><summary><s>Steps(simplified) taken to update the dependency:</s></summary>
<p>

1. Update Go version to `1.20`.
- Change all the non-k8s pkg version to `latest`
- core k8s related pkg version to `0.28.6`(https://github.com/kubernetes/api/releases/tag/v0.28.6, https://github.com/kubernetes/apimachinery/releases/tag/v0.28.6)
- rancher related k8s pkg version to `v1.28.6-rancher1`(https://github.com/rancher/client-go/releases/tag/v1.28.6-rancher1)
- rancher/rancher to the HEAD commit of branch [release/v2.8.1](https://github.com/rancher/rancher/tree/release/v2.8.1) (we might need to update to HEAD commit of branch [release/v2.9](https://github.com/rancher/rancher/tree/release/v2.9/) soon).

```
module github.com/rancher/hosted-providers-e2e

go 1.20

require (
	github.com/Masterminds/semver/v3 latest
	github.com/epinio/epinio latest
	github.com/onsi/ginkgo/v2 latest
	github.com/onsi/gomega latest
	github.com/pkg/errors latest
	github.com/rancher/rancher fa9e804ff2d7 // HEAD commit of release/v2.8.1 branch
	k8s.io/apimachinery v0.28.6    // to add support of k8s 1.28, upgrade all the related pkgs to 0.28.6
)

replace (
	k8s.io/api => k8s.io/api v0.28.6
	k8s.io/client-go => github.com/rancher/client-go v1.28.6-rancher1 // to add support of k8s 1.28, upgrade this to 1.28.6-rancher1
)

replace github.com/rancher/rancher/pkg/client => github.com/rancher/rancher/pkg/client v0.0.0-20231121113554-764f891cfa95

replace github.com/rancher/rancher/pkg/apis => github.com/rancher/rancher/pkg/apis v0.0.0-20231121113554-764f891cfa95
```
2. `go1.20.3 mod download`
```
module github.com/rancher/hosted-providers-e2e

go 1.20

require (
	github.com/Masterminds/semver/v3 v3.2.1
	github.com/epinio/epinio v1.11.0
	github.com/onsi/ginkgo/v2 v2.15.0
	github.com/onsi/gomega v1.31.1
	github.com/pkg/errors v0.9.1
	github.com/rancher/rancher v0.0.0-20240118221102-fa9e804ff2d7
	k8s.io/apimachinery v0.28.6
)

replace (
	k8s.io/api => k8s.io/api v0.28.6
	k8s.io/client-go => github.com/rancher/client-go v1.28.6-rancher1
)

replace github.com/rancher/rancher/pkg/client => github.com/rancher/rancher/pkg/client v0.0.0-20231121113554-764f891cfa95

replace github.com/rancher/rancher/pkg/apis => github.com/rancher/rancher/pkg/apis v0.0.0-20231121113554-764f891cfa95
```
3. Update `rancher/rancher/apis` and `rancher/rancher/client` pkgs.
```
$ go1.20.13 mod edit -replace=github.com/rancher/rancher/pkg/apis=github.com/rancher/rancher/pkg/apis@v0.0.0-20240118221102-fa9e804ff2d7

$ go1.20.13 mod edit -replace=github.com/rancher/rancher/pkg/client=github.com/rancher/rancher/pkg/client@v0.0.0-20240118221102-fa9e804ff2d7

```

```
module github.com/rancher/hosted-providers-e2e

go 1.20

require (
	github.com/Masterminds/semver/v3 v3.2.1
	github.com/epinio/epinio v1.11.0
	github.com/onsi/ginkgo/v2 v2.15.0
	github.com/onsi/gomega v1.31.1
	github.com/pkg/errors v0.9.1
	github.com/rancher/rancher v0.0.0-20240118221102-fa9e804ff2d7
	k8s.io/apimachinery v0.28.6
)

replace (
	k8s.io/api => k8s.io/api v0.28.6
	k8s.io/client-go => github.com/rancher/client-go v1.28.6-rancher1
)

replace github.com/rancher/rancher/pkg/client => github.com/rancher/rancher/pkg/client v0.0.0-20240118221102-fa9e804ff2d7

replace github.com/rancher/rancher/pkg/apis => github.com/rancher/rancher/pkg/apis v0.0.0-20240118221102-fa9e804ff2d7

```
4. `go1.20.13 mod tidy`
6. `go1.20.13 mod vendor`

</p>
</details> 

<details><summary>Steps taken to update the dependencies:</summary>
<p>

1. Update Go version to `1.20`.
- Change all the non-k8s pkg version to `latest`
- core k8s related pkg version to `0.28.6`(https://github.com/kubernetes/api/releases/tag/v0.28.6, https://github.com/kubernetes/apimachinery/releases/tag/v0.28.6)
- rancher related k8s pkg version to `v1.28.6-rancher1`(https://github.com/rancher/client-go/releases/tag/v1.28.6-rancher1)
- move from rancher/rancher to rancher/shepherd; using the HEAD commit of main branch.

```go
module github.com/rancher/hosted-providers-e2e

go 1.20

require (
    github.com/onsi/ginkgo/v2 latest
    github.com/onsi/gomega latest
    github.com/epinio/epinio latest
    github.com/rancher/shepherd 3690b7e         // release/v2.8.1 HEAD commit
    github.com/stretchr/testify latest
    k8s.io/api v0.27.9
    k8s.io/apimachinery v0.27.9
)

replace (
    k8s.io/api => k8s.io/api v0.27.9
    k8s.io/client-go => github.com/rancher/client-go v1.27.4-rancher1
)
```
2. `go mod tidy`
3. `go mod vendor`

</p>
</details>

cpinjani commented on go.mod:
```diff
@@ -1,16 +1,16 @@
-	github.com/rancher/rancher v0.0.0-20231121113554-764f891cfa95
-	k8s.io/apimachinery v0.27.6
-	k8s.io/utils v0.0.0-20230505201702-9f6742963106
+	github.com/rancher/rancher v0.0.0-20240118221102-fa9e804ff2d7
```
lgtm.
Is this commit from v2.8.1 ? We should take latest commit from v2.8 branch instead

cpinjani commented on docs/module_dep.md:
```diff
@@ -1,84 +1,42 @@
+    github.com/onsi/ginkgo/v2 latest
+    github.com/onsi/gomega latest
+    github.com/epinio/epinio latest
+    github.com/rancher/shepherd 3690b7e         // release/v2.8.1 HEAD commit
```
```suggestion
    github.com/rancher/shepherd 3690b7e         // shepherd main HEAD commit
```

valaparthvi commented:
@cpinjani I have rebased the branch on top of main and also reverted some changes related to upgrading k8s related pkgs. `rancher/shepherd` is still on k8s.io/api and k8s.io/apimachinery and k8s.io/client-go v0.27.9, they will upgrade once `rancher/rancher` does, so we will do it once they do.

cpinjani commented:
> @cpinjani I have rebased the branch on top of main and also reverted some changes related to upgrading k8s related pkgs. `rancher/shepherd` is still on k8s.io/api and k8s.io/apimachinery and k8s.io/client-go v0.27.9, they will upgrade once `rancher/rancher` does, so we will do it once they do.

Sure, are the tests run and passing on these latest changes ?

valaparthvi commented:
> > @cpinjani I have rebased the branch on top of main and also reverted some changes related to upgrading k8s related pkgs. `rancher/shepherd` is still on k8s.io/api and k8s.io/apimachinery and k8s.io/client-go v0.27.9, they will upgrade once `rancher/rancher` does, so we will do it once they do.
> 
> Sure, are the tests run and passing on these latest changes ?

Yes, check the PR description, the 4th runs have passed.

cpinjani reviewed (approved):
lgtm
